[GH-160]  Credo and Dialyxer fixes

untra opened the pull request:
* [original PR](#180)
* Fixes all remaining credo errors
* Fixes many dializer issues
* Enforces a `@type t` typespec for `Block`, `Header`, `TxData`, `SignedTx`, etc. ensuring consistency
* Enforces consistent parens on all types, _erlang-style_ (eg. `integer()`)
* Structs are typed with a t (eg. `Block.t()`)
* Refactored a few functions to be more readable

This closes #160.
This closes #161.

untra commented:
:exclamation: Please take a look over this PR. I've squashed my commits and changes, and would like to get to a good stopping point with this PR before additional changes to this repo are merged in.

There are some Dialyzer errors that are false positives. A few anonymous functions without a return type are flagged by Dialyzer as:
> The created fun has no local return
`no local return` is a common dialyzer error, but it seems to be a weird interaction in an elixir environment that causes it to report this. No real fix.

There are still a few other Dialyzer errors I want to triple-check, but it's getting late, and I want eyes :eyes: on this PR, as it touches a lot of files.

cytadela8 commented on apps/aecore/lib/aecore/chain/block_validation.ex:
```diff
@@ -8,7 +8,7 @@ defmodule Aecore.Chain.BlockValidation do
   alias Aecore.Chain.Difficulty
 
-  @spec validate_block!(Block.block(), Block.block(), map(), list()) :: {:error, term()} | :ok
+  @spec validate_block!(%Block{}, %Block{}, map(), list(%Block{})) :: {:error, term()} | :ok
```
Shouldn't we use Block.t()? (add `@type t :: block()` to `Aecore.Structures.Block`  and use it)

untra replied:
`Aecore.Structures.Block` does have @type t :: %Block{} [currently](https://github.com/aeternity/elixir-research/pull/182/commits/426bab6bd5887316433c961ecba43f9c985dbc86#diff-ba597a8315c1932793bd18d5cb93f2aeR9) .

A previous comment [here](https://github.com/aeternity/elixir-research/pull/180#discussion_r159038716) suggested that `%Block{}` would be preferable to `Block.t()` . (I like `Block.t` but without the erlang-style parens). I have changed all the  `Aecore.Structures` to using the `%Block{}` style. The team does prefer erlang-style typings, so the endings with `Block.t()` will have parens. The correct typespec style is entirely a team preference. So I'll ask the question again to be sure:

:question: Which style does the team prefer; `%Block{}` or `Block.t()`?

thepiwo replied:
lets go with `Block.t()`

cytadela8 commented on apps/aecore/lib/aecore/chain/difficulty.ex:
```diff
@@ -40,7 +41,7 @@ defmodule Aecore.Chain.Difficulty do
   end
 
-  @spec calculate_distance(list) :: float()
+  @spec calculate_distance(list()) :: float()
```
`@spec calculate_distance(list(%Block{})) :: float()`

cytadela8 commented on apps/aecore/lib/aecore/peers/worker.ex:
```diff
@@ -244,6 +240,7 @@ defmodule Aecore.Peers.Worker do
     end
   end
 
+  # TODO: this function is never called, should it be removed?
```
It's called by `add_peer`.

cytadela8 replied:
line 222

cytadela8 commented on apps/aecore/lib/aecore/peers/worker.ex:
```diff
@@ -258,28 +255,27 @@ defmodule Aecore.Peers.Worker do
     :ets.new(:nonce_table, [:named_table])
   end
 
+  # TODO: this function is never called, should it be removed?
```
It's called by `add_peer`.

cytadela8 replied:
line 215

cytadela8 commented on apps/aecore/lib/aecore/peers/worker.ex:
```diff
@@ -258,28 +255,27 @@ defmodule Aecore.Peers.Worker do
   end
 
-  defp should_a_peer_be_added peers_count do
+  # TODO: this function is never called, should it be removed?
```
It's called by `add_peer` line 211.

thepiwo commented:
> Structs are typed like their use (eg. %Block{})

--> Structs are typed like their use (eg. Block.t() or list(Block.t())

untra commented:
@thepiwo @cytadela8 
How does it look now? :+1:

thepiwo commented:
Thank you, looks great!
In the def of some functions we use something like `%Block{} = block` or `%Header{nonce: nonce} = block_header`. Is this type checks on runtime and best practice to do?

If so should we refactor this to be `Block.t() = block` as well?

thepiwo commented:
`@type t :: %Block{} | %Block{header: Header.t(), txs: list(TxData.t())}` the pipe means eighter this or that? So I think we should exclude the empty type definitions like `%Block{}` from the specification, to prevent bugs.

cytadela8 commented:
I think we can merge now.

thepiwo commented:
@untra should we merge?

untra commented:
@thepiwo I addressed the `@type t :: %Block{} | %Block{header: Header.t(), txs: list(TxData.t())}` detail on the structs, and fixed the conflicting files. Give it one more look over, and this PR is ready for a merge :+1:

thepiwo commented:
@meivantodorov can you approve and merge?

untra commented:
Is this ready for a merge? I think this PR should get in sooner rather than later 🙏
